src/bvals/bvals_swarm.cpp

```diff
@@ -0,0 +1,166 @@
+    bd.flag[n] = BoundaryStatus::waiting;
+#ifdef MPI_PARALLEL
+    bd.req_send[n] = MPI_REQUEST_NULL;
+    bd.req_recv[n] = MPI_REQUEST_NULL;
+#endif
+  }
+}
+
+void BoundarySwarm::SetupPersistentMPI() {
+#ifdef MPI_PARALLEL
+  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
+  int &mylevel = pmb->loc.level;
+
+  // Initialize neighbor communications to other ranks
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+
+    // Neighbor on different MPI process
+    if (nb.snb.rank != Globals::my_rank) {
+      send_tag[n] = pmb->pbval->CreateBvalsMPITag(nb.snb.lid, nb.targetid, swarm_id_);
+      recv_tag[n] = pmb->pbval->CreateBvalsMPITag(pmb->lid, nb.bufid, swarm_id_);
+      if (bd_var_.req_send[nb.bufid] != MPI_REQUEST_NULL) {
+        MPI_Request_free(&bd_var_.req_send[nb.bufid]);
+      }
+      if (bd_var_.req_recv[nb.bufid] != MPI_REQUEST_NULL) {
+        MPI_Request_free(&bd_var_.req_recv[nb.bufid]);
+      }
+    }
+  }
+#endif
+}
+
+// Send particle buffers across meshblocks. If different MPI ranks, use MPI, if same rank,
+// do a deep copy on device.
+void BoundarySwarm::Send(BoundaryCommSubset phase) {
+  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
+  int &mylevel = pmb->loc.level;
+  // Fence to make sure buffers are loaded before sending
+  pmb->exec_space.fence();
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+    if (nb.snb.rank != Globals::my_rank) {
+#ifdef MPI_PARALLEL
+      PARTHENON_REQUIRE(bd_var_.req_send[nb.bufid] == MPI_REQUEST_NULL,
+                        "Trying to create a new send before previous send completes!");
+      MPI_Isend(bd_var_.send[n].data(), send_size[n], MPI_PARTHENON_REAL, nb.snb.rank,
+                send_tag[n], MPI_COMM_WORLD, &(bd_var_.req_send[nb.bufid]));
+#endif // MPI_PARALLEL
+    } else {
+      MeshBlock &target_block = *pmy_mesh_->FindMeshBlock(nb.snb.gid);
+      std::shared_ptr<BoundarySwarm> ptarget_bswarm =
+          target_block.pbswarm->bswarms[bswarm_index];
+      if (send_size[nb.bufid] > 0) {
+        // Ensure target buffer is large enough
+        if (bd_var_.send[nb.bufid].extent(0) >
+            ptarget_bswarm->bd_var_.recv[nb.targetid].extent(0)) {
+          ptarget_bswarm->bd_var_.recv[nb.targetid] =
+              ParArray1D<Real>("Buffer", (bd_var_.send[nb.bufid].extent(0)));
+        }
+
+        target_block.deep_copy(ptarget_bswarm->bd_var_.recv[nb.targetid],
+                               bd_var_.send[nb.bufid]);
+        ptarget_bswarm->recv_size[nb.targetid] = send_size[nb.bufid];
+        ptarget_bswarm->bd_var_.flag[nb.targetid] = BoundaryStatus::arrived;
+      } else {
+        ptarget_bswarm->recv_size[nb.targetid] = 0;
+        ptarget_bswarm->bd_var_.flag[nb.targetid] = BoundaryStatus::completed;
+      }
+    }
+  }
+}
+
+void BoundarySwarm::Receive(BoundaryCommSubset phase) {
+#ifdef MPI_PARALLEL
+  MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
Can you add some comments explaining why this barrier is necessary?

Reply:
Oops, these were statements I used for debugging that I forgot to remove. They are unnecessary (and explode our whole model for god parallel scaling). Thanks for noticing them, they're gone now.